src/intptrcast.rs

```diff
@@ -0,0 +1,103 @@
+            int_to_ptr_map: Vec::default(),
+            next_base_addr: 2u64.pow(16)
+        }
+    }
+}
+
+impl<'mir, 'tcx> GlobalState {
+    pub fn int_to_ptr(
+        base_addr: u64,
+        memory: &Memory<'mir, 'tcx, Evaluator<'tcx>>,
+    ) -> InterpResult<'tcx, Pointer<Tag>> {
+        let global_state = memory.extra.intptrcast.borrow();
+        
+        match global_state.int_to_ptr_map.binary_search_by_key(&base_addr, |(addr, _)| *addr) {
+            Ok(pos) => {
+                let (_, alloc_id) = global_state.int_to_ptr_map[pos];
+                // `base_addr` is the starting address for an allocation, the offset should be
+                // zero. The pointer is untagged because it was created from a cast
+                Ok(Pointer::new_with_tag(alloc_id, Size::from_bytes(0), Tag::Untagged))
+            },
+            Err(0) => err!(DanglingPointerDeref), 
+            Err(pos) => {
+                // This is the gargest of the adresses smaller than `base_addr`,
+                // i.e. the greatest lower bound (glb)
+                let (glb, alloc_id) = global_state.int_to_ptr_map[pos - 1];
+                // This never overflows because `base_addr >= glb`
+                let offset = base_addr - glb;
+                // If the offset exceeds the size of the allocation, this access is illegal
+                if offset <= memory.get(alloc_id)?.bytes.len() as u64 {
+                    // This pointer is untagged because it was created from a cast
+                    Ok(Pointer::new_with_tag(alloc_id, Size::from_bytes(offset), Tag::Untagged))
+                } else {
+                    err!(DanglingPointerDeref)
+                } 
+            }
+        }
+    }
+
+    pub fn ptr_to_int(
+        ptr: Pointer<Tag>,
+        memory: &Memory<'mir, 'tcx, Evaluator<'tcx>>,
+    ) -> InterpResult<'tcx, u64> {
+        let mut global_state = memory.extra.intptrcast.borrow_mut();
+
+        let alloc = memory.get(ptr.alloc_id)?;
+
+        let base_addr = match alloc.extra.intptrcast.base_addr.get() { 
+            Some(base_addr) => base_addr,
+            None => {
+                let base_addr = Self::align_addr(global_state.next_base_addr, alloc.align.bytes());
+
+                global_state.next_base_addr = base_addr + alloc.bytes.len() as u64;
+                alloc.extra.intptrcast.base_addr.set(Some(base_addr));
+
+                let elem = (base_addr, ptr.alloc_id);
+
+                // Given that `next_base_addr` increases in each allocation, pushing the
+                // corresponding tuple keeps `int_to_ptr_map` sorted
+                global_state.int_to_ptr_map.push(elem); 
+
+                base_addr
+            }
+        };
+
+        Ok(base_addr + ptr.offset.bytes())
+    }
+
+    /// Shifts `addr` to keep it aligned with `align` by rounding up `addr` to the next multiple of
+    /// `align`
+    fn align_addr(addr: u64, align: u64) -> u64 {
+        addr + align - addr % align
```

Comment:
This does not match what the comment says: if `addr % align == 0`, "rounding up to next multiple" would return `addr`, but you are returning `addr + align`. That's fine for the allocator, but you should make sure comment and implementation match.

Reply:
Why did you mark this as resolved? Comment and code still don't seem to match?

Reply:
Oh sorry, I usually mark the changes as resolved when checking I've done everything just before doing push

Reply:
But, uh, you didn't actually do this? Or am I missing something?

> This does not match what the comment says: if addr % align == 0, "rounding up to next multiple" would return addr, but you are returning addr + align. That's fine for the allocator, but you should make sure comment and implementation match.

Reply:
This is probably my fault by force pushing too much (but I'm scared of rebases right now) 
https://github.com/rust-lang/miri/pull/779/files#diff-550368cb20a0eeb68d2e64ffb75c0c50R95
```
    /// Shifts `addr` to make it aligned with `align` by rounding `addr` to the smallest multiple
    /// of `align` that is larger or equal to `addr`
```

Reply:
That's not what the code does. The code rounds up to the next multiple that's *strictly larger* than `addr`. Or am I misreading?

Reply:
No, it is correct.

Reply:
"it" = the code or "it" = my understanding?

Reply:
your understanding. The code is fixed now

Reply:
did you forget to push? :D

Reply:
That sounds completely possible, but no, I pushed. It is being tested by travis now.

Reply:
The code is entirely unchanged as far as I can tell

Reply:
https://github.com/rust-lang/miri/compare/91609bf353a503c30d4f6b4e0fa7eb08f45a5aac..7fbf8e53a22925a5ff865b3546f881f7542c4f9e

Reply:
ah lol, I thought the code was getting changed :D Open an issue so we stop wasting `align` bytes if we were already aligned?

Reply:
We already have an issue to tweak our alignment logic, just add it there.